token/token.go

```diff
@@ -0,0 +1,67 @@
+package token
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/subtle"
+	"errors"
+	"fmt"
+)
+
+var (
+	// Error indicating that a token was malformed.
+	ErrInvalidToken = errors.New("Invalid token.")
+
+	// Indicating that verification failed, because the token was
+	// incorrect.
+	ErrIncorrectToken = errors.New("Incorrect token.")
+)
+
+// A cryptographically random 128-bit value.
+type Token [128 / 8]byte
+
+// Generate a new, cryptographically random token.
+func New() (Token, error) {
+	var ret Token
+	_, err := rand.Read(ret[:])
+	return ret, err
+}
+
+func (t Token) MarshalText() ([]byte, error) {
+	return []byte(fmt.Sprintf("%0x", t)), nil
+}
+
+func (t *Token) UnmarshalText(text []byte) error {
+	if len(text) != 2*len(t[:]) {
```

Comment:
Can you remind me why it's `2*len`?

Reply:
We're decoding bytes from hex, so each byte is two chars.